File: PnL/src/pandas/merge RCL.py
import pandas as pd
from pathlib import Path
from janitor import clean_names


# Path
path = Path(__file__).parent.parent.parent


# Functions
def read_multiple_files(list_of_files):
    dataframes = [
        pd.read_excel(
            file,
            sheet_name="RCL Input GC",
            skiprows=4,
            dtype={"Item Structure": str, "Outlet": str, "Plant": str},
        )
        for file in list_of_files
    ]

    # Add a new column with filename to each DataFrame
    for i, df in enumerate(dataframes):
        df["source"] = list_of_files[i].stem

    # Merge the list of DataFrames into a single DataFrame
    df = pd.concat(dataframes)

    # reorder columns
    df = df[["source"] + [col for col in df.columns if col not in ["source"]]]

    return df


def rename_columns(df):
    df = df.drop(columns=df.loc[:, "ACT Period GC":"FC GC fx impact"])
    df = df.rename(
        columns={
            "Single Period BUD GC": "Period Plan",
            "Single Period ACT GC @ BUD fx": "Period Act",
            "Single Period ACT-BUD GC @ BUD fx": "Delta Period",
            "Comment Single Period ACT-BUD Period": "Comments Period",
            "YTD BUD GC": "YTD Plan",
            "YTD ACT GC @ BUD fx": "YTD Act",
            "YTD ACT-BUD GC @ BUD fx": "Delta YTD",
            "Comment YTD ACT-BUD": "Comments YTD",
            "FY BUD GC": "Plan",
            "FC GC @ BUD fx": "FC",
            "FC - BUD GC @ BUD fx": "Delta to Plan",
            "Comment FC - BUD GC": "Comments FC",
            "LFC GC @ BUD fx": "LFC",
            "FC GC @ BUD fx.1": "FC_",
            "FC - LFC GC @ BUD fx": "Delta to LFC",
            "Comment FC - LFC": "Comments FC changes",
        }
    )
    return df


def convert_currency(df, bud_fx):
    # Change GC to LC
    # DataFrame.select_dtypes method
    df.loc[:, df.select_dtypes(include=[float]).columns] *= bud_fx
    return df


def extract_poc(df):
    # Get Outlet, Plant information from file names using Regex
    df["outlet_plant"] = df["source"].str.extract(r"([0-9\_]{7,9})")
    df["outlet"] = df["outlet_plant"].str.extract(r"([0-9]{3,4})")
    df["plant"] = df["outlet_plant"].str.extract(r"([0-9]{3}$)")
    # Remove columns
    df = df.drop(columns=["source", "outlet_plant"])
    return df


def poc(filename):
    df = pd.read_csv(filename, dtype="str").clean_names()
    df = df.drop(columns=["cu", "profit_center"])
    return df


def join_with_poc(df, filename):
    # Plant Outlet Combination
    poc_df = poc(filename)
    df = df.merge(poc_df, on=["outlet", "plant"], how="left")
    # Reorder columns
    head_col = ["division", "bu", "outlet_name", "plant_name", "outlet", "plant"]
    df = df[head_col + [col for col in df.columns if col not in head_col]]
    return df


def add_key_column(df):
    # Add key column for look up RCL comments
    df["key"] = df["outlet"] + "_" + df["plant"] + "_" + df["rcl_item_structure"]
    return df


def main():

    # Path
    data_path = path / "data" / "RCL"

    # Variables
    bud_fx = 1329  # Budget FX rate in 2023

    # Filenames
    meta_file = path / "meta" / "POC.csv"
    output_file = path / "output" / "RCL.csv"

    # Input data: List of multiple text files
    xls_files = [
        file
        for file in data_path.iterdir()
        if file.is_file() and file.suffix == ".xlsm"
    ]

    # Read data
    df = read_multiple_files(xls_files)

    # Process data
    df = (
        df.pipe(rename_columns)
        .pipe(clean_names)
        .pipe(convert_currency, bud_fx)
        .pipe(extract_poc)
        .pipe(join_with_poc, meta_file)
        .pipe(add_key_column)
    )

    # Write data
    df.to_csv(output_file, index=False)
    print("A file is created")


if __name__ == "__main__":
    main()

File: PnL/src/dplyr/merge RCL.R
library(dplyr)
library(readr)
library(purrr)
library(stringr)
library(readxl)
library(fs)
library(janitor)
library(here)


# Path
path <- here("PnL")


# Functions
read_multiple_files <- function(list_of_files) {
  df <- list_of_files |>
    map_dfr(read_excel,
      sheet = "RCL Input GC", skip = 4, .id = "source",
      col_types = c(
        "text", "numeric", "numeric", "numeric",
        "text", "numeric", "numeric", "numeric",
        "text", "numeric", "numeric", "numeric",
        "text", "numeric", "numeric", "numeric",
        "text", "numeric", "numeric", "numeric",
        "numeric", "numeric", "numeric"
      ),
    )
  return(df)
}

rename_columns <- function(df) {
  df <- df |>
    select(!c("ACT Period GC":"FC GC fx impact")) |>
    rename(c(
      `Period Plan`         = "Single Period BUD GC",
      `Period Act`          = "Single Period ACT GC @ BUD fx",
      `Delta Period`        = "Single Period ACT-BUD GC @ BUD fx",
      `Comments Period`     = "Comment Single Period ACT-BUD Period",
      `YTD Plan`            = "YTD BUD GC",
      `YTD Act`             = "YTD ACT GC @ BUD fx",
      `Delta YTD`           = "YTD ACT-BUD GC @ BUD fx",
      `Comments YTD`        = "Comment YTD ACT-BUD",
      `Plan`                = "FY BUD GC",
      `FC`                  = "FC GC @ BUD fx...11",
      `Delta to Plan`       = "FC - BUD GC @ BUD fx",
      `Comments FC`         = "Comment FC - BUD GC",
      `LFC`                 = "LFC GC @ BUD fx",
      `FC_`                 = "FC GC @ BUD fx...15",
      `Delta to LFC`        = "FC - LFC GC @ BUD fx",
      `Comments FC changes` = "Comment FC - LFC"
    ))
  return(df)
}

convert_currency <- function(df, bud_fx) {
  # Change GC to LC
  df <- df |>
    mutate(across(where(is.numeric), ~ .x * bud_fx))
  return(df)
}

extract_poc <- function(df) {
  # Get Outlet, Plant infomration from file names using Regex
  df <- df |>
    mutate(
      outlet_plant = str_extract(source, "[0-9\\_]{7,9}"),
      outlet = str_extract(df$outlet_plant, "[0-9]{3,4}"),
      plant = str_extract(df$outlet_plant, "[0-9]{3}$")
    ) |>
    select(!c("source", "outlet_plant")) |>
    relocate(c("outlet", "plant"))
  return(df)
}

# Plant Outlet Combination
poc <- function(filename) {
  df <- read_csv(filename, col_types = "ccc") |> # "c" as Character type
    clean_names() |>
    select(!c("cu", "profit_center"))
  return(df)
}

join_with_poc <- function(df, filename) {
  df <- df |>
    left_join(poc(filename), by = c("outlet", "plant")) |>
    relocate(c("division", "bu", "outlet_name", "plant_name"))
  return(df)
}

# Add key column for look up RCL comments
add_key_column <- function(df) {
  df <- df |>
    tidyr::unite("key", c("outlet", "plant", "rcl_item_structure"),
      remove = FALSE
    ) |>
    relocate("key", .after = last_col())
  return(df)
}

main <- function() {
  # Variables
  bud_fx <- 1329 # Budget FX rate in 2023

  # Path
  data_path <- here(path, "data", "RCL")

  # Filenames
  xls_files <- dir_ls(data_path, regexp = "\\.xlsm$")
  meta_file <- here(path, "meta", "POC.csv")
  output_file <- here(path, "output", "RCL.csv")


  # Read data
  df <- read_multiple_files(xls_files)

  # Process data
  df <- df |>
    rename_columns() |>
    clean_names() |>
    convert_currency(bud_fx) |>
    extract_poc() |>
    join_with_poc(meta_file) |>
    add_key_column()


  # Write data
  write_csv(df, output_file, na = "")
}

main()
